### ISSUE TITLE
how to commit changes if checks fail (without using PR review suggestions)

### ISSUE THREAD

**Issue by 2bndy5** [cpp-linter/cpp-linter#82]
We should probably have this as an optional feature since our runtime has all the tools it needs to do it (provided the proper permissions in CI).

**Comment by 2bndy5** [comment]
> Do you mean how to commit changes made by cpp-linter after execution (e.g. create a PR

Yes, but it makes more sense to push to the branch that is checked out, not open a new PR. The [pre-commit.ci app does the same thing](https://pre-commit.ci/#features).

**Comment by 2bndy5** [comment]
The actual problem that this issue solves is the fact that cpp-linter-action uses its own venv that the user's CI workflow cannot use once cpp-linter-action step is finished.

> Or users can DIY their own using git commands

I like this idea. The `auto-fix` option would just have to leave the file changes in the workspace directory. Then the user can `git add --all` (or use gh-cli) and push the changes under whatever authorization they want.

So, maybe we could have 2 options:
- `fix-worktree`: leave clang-tool changes applied to workspace
- `push-fixes`: push any changes from clang tools

Each idea would require cpp-linter to remove any temp files (`.cpp-linter_cache/*`) before exiting.

Considering the extra configurations around auto-pushing a commit (like commit message, sign-off trailers, etc), it might just be easier to only implement `fix-worktree`.

---
Given that this pkg runs both clang-tidy and clang-format, it is worth noting that changes should be applied in the following order:

1. clang-tidy changes (if desired)
2. clang-format changes

That should make sure that the changes from each tool don't conflict with each other.

**Comment by 2bndy5** [comment]
I've been thinking about this feature recently... Given the agreed upon workflow:
```mermaid
flowchart LR
    step1[apply clang-tidy changes to a cached copy of files]
    step2[apply clang-format changes to cache clang-tidy changes]
    step3[copy cache back into worktree]

    step1 --> step2 --> step3
```

It would be very difficult to determine which lines were changed in the PR after applying clang-tidy changes. Telling clang-format which lines to scan/fix (of the cached copy) might require a 3-way diff. I'm not sure how to do that. 

# TL;DR
It might be easier to allow only committing clang-format today changes and optionally running clang format on just the clang-tidy changes (which may not be all the changes in the PR).

**Comment by 2bndy5** [comment]
>> Or users can DIY their own using git commands
>
> I like this idea. The auto-fix option would just have to leave the file changes in the workspace directory. Then the user can git add --all and push the changes under whatever authorization they want.

I'm working toward this over at cpp-linter-rs. I know we settled on a new `push-commit` type option, but I think an artifact would be a better fit. I'm envisioning a single patch file that users could `git apply` or do whatever they want. For example:

```yml
- name: run cpp-linter
  id: linter
  with:
    auto-fix: true
    # ... other desired inputs
  # let a new output named `patch` point to the resulting patch file.
- name: upload patch file as artifact
  if: ${{ steps.linter.outputs.patch }} # only if a patch is made
  uses: actions/upload-artifact@v6
  with:
    # by default, name of patch is "cpp-linter.patch"
    path: ${{ steps.linter.outputs.patch }}
    name: cpp-linter-fixes
 
# ... In a separate job to keep the permissive token use limited
- name: download patch
  uses: actions/download-artifact@v6
  with:
    name: cpp-linter-fixes
- name: Apply fixes
  run: |-
    git apply cpp-linter.patch # could also add filtering args to git apply
    git commit -m "whatever the user decides as a message"
    git push
    # (optional) exit non-zero to show CI failure in commit history.
    exit 1
```
